src/_tkagg.cpp

```diff
@@ -95,8 +96,101 @@ static PyObject *mpl_tk_blit(PyObject *self, PyObject *args)
     }
 }
 
+#ifdef WIN32_DLL
+LRESULT CALLBACK
+DpiSubclassProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam,
+                UINT_PTR uIdSubclass, DWORD_PTR dwRefData)
+{
+    switch (uMsg) {
+    case WM_DPICHANGED:
+        // This function is a subclassed window procedure, and so is run during
+        // the Tcl/Tk event loop. Thus, we must re-take the GIL. Unfortunately,
+        // Tkinter has a *second* lock on Tcl threading that is not exposed
+        // publicly, so we must not call *any* Tk code from here, thus need to
+        // use a polled queue instead.
+        {
+            PyObject *queue = (PyObject *)dwRefData;
+            PyGILState_STATE state = PyGILState_Ensure();
+            PyObject_CallMethod(queue, "put", "i", LOWORD(wParam));
+            PyGILState_Release(state);
+        }
+        return 0;
+    case WM_NCDESTROY:
+        RemoveWindowSubclass(hwnd, DpiSubclassProc, uIdSubclass);
+        break;
+    }
+
+    return DefSubclassProc(hwnd, uMsg, wParam, lParam);
+}
+#endif
+
+static PyObject*
+mpl_tk_enable_dpi_awareness(PyObject* self, PyObject*const* args,
```

Comment:
should this live in backend_tk, or does it rather belong in _c_internal_utils? (I don't know...)

Reply:
I put the DPI awareness setup in `_c_internal_utils` because it is not backend-specific (theoretically, a toolkit could support DPI changes, but expect the application to enable it with the API or a manifest, but I have not really checked this). This function is specific to Tk though, and may or may not be portable to other toolkits.